File: priv/javascript/lib/kernel.js
import SpecialForms from './kernel/special_forms';
import Patterns from './patterns/patterns';
import Tuple from './tuple';
import BitString from './bit_string';
import { Protocol } from './protocol';
import { PID } from './processes/processes';

function tl(list){
  return SpecialForms.list(...list.slice(1));
}

function hd(list){
  return list[0];
}

function is_nil(x){
  return x === null;
}

function is_atom(x){
  return typeof x === 'symbol';
}

function is_binary(x){
  return typeof x === 'string' || x instanceof String;
}

function is_boolean(x){
  return typeof x === 'boolean' || x instanceof Boolean;
}

function is_function(x, arity = -1){
  return typeof x === 'function' || x instanceof Function;
}

function is_float(x){
  return is_number(x) && !Number.isInteger(x);
}

function is_integer(x){
  return Number.isInteger(x);
}

function is_list(x){
  return x instanceof Array;
}

function is_map(x){
  return typeof x === 'object' || x instanceof Object;
}

function is_number(x){
  return typeof x === "number";
}

function is_tuple(x){
  return x instanceof Tuple;
}

function length(x){
  return x.length;
}

function is_pid(x){
  return x instanceof PID;
}

function is_port(x){
  return false;
}

function is_reference(x){
  return false;
}

function is_bitstring(x){
  return is_binary(x) || x instanceof BitString;
}

function __in__(left, right){
  for(let x of right){
    if(match__qmark__(left, x)){
      return true;
    }
  }

  return false;
}

function abs(number){
  return Math.abs(number);
}

function round(number){
  return Math.round(number);
}

function elem(tuple, index){
  if(is_list(tuple)){
    return tuple[index];
  }

  return tuple.get(index);
}

function rem(left, right){
  return left % right;
}

function div(left, right){
  return left / right;
}

function and(left, right){
  return left && right;
}

function or(left, right){
  return left || right;
}

function not(arg){
  return !arg;
}

function apply(...args){
  if(args.length === 3){
    let mod = args[0];
    let func = args[1];
    let func_args = args[2];
    return mod[func].apply(null, func_args);
  }else{
    let func = args[0];
    let func_args = args[1];

    return func.apply(null, func_args);
  }
}

function to_string(arg){
  if(is_tuple(arg)){
    return Tuple.to_string(arg);
  }

  return arg.toString();
}

function match__qmark__(pattern, expr, guard = () => true){
  return Patterns.match_no_throw(pattern, expr, guard) != null;
}

function defstruct(defaults){
  return class {
    constructor(update = {}){
      let the_values = Object.assign(defaults, update);
      Object.assign(this, the_values);
    }

    static create(updates = {}){
      let x = new this(updates);
      return Object.freeze(x);
    }
  }
}


function defexception(defaults){
  return class extends Error {
    constructor(update = {}){
      let message = update.message || "";
      super(message);

      let the_values = Object.assign(defaults, update);
      Object.assign(this, the_values);

      this.name = this.constructor.name;
      this.message = message;
      this[SpecialForms.atom("__exception__")] = true;
      Error.captureStackTrace(this, this.constructor.name);
    }

    static create(updates = {}){
      let x = new this(updates);
      return Object.freeze(x);
    }
  }
}

function defprotocol(spec){
  return new Protocol(spec);
}

function defimpl(protocol, type, impl){
  protocol.implementation(type, impl);
}

export default {
  SpecialForms,
  tl,
  hd,
  is_nil,
  is_atom,
  is_binary,
  is_boolean,
  is_function,
  is_float,
  is_integer,
  is_list,
  is_map,
  is_number,
  is_tuple,
  length,
  is_pid,
  is_port,
  is_reference,
  is_bitstring,
  in: __in__,
  abs,
  round,
  elem,
  rem,
  div,
  and,
  or,
  not,
  apply,
  to_string,
  match__qmark__,
  defstruct,
  defprotocol,
  defimpl
};

File: lib/elixir_script/lib/kernel.ex
defmodule ElixirScript.Kernel do

  def abs(number) do
    Math.abs(number)
  end

  def apply(fun, args) do
    fun.apply(nil, args)
  end

  def apply(module, fun, args) do
    module[fun].apply(nil, args)
  end

  def binary_part(binary, start, length) do
    binary.substring(start, length)
  end

  def div(left, right) do
    left / right
  end

  def hd(list) do
    list[0]
  end

  def tl(list) do
    list.slice(1)
  end

  defmacro is_atom(term) do
    quote do
      typeof(unquote(term)) === "symbol"
    end
  end

  defmacro is_binary(term) do
    quote do
      typeof(unquote(term)) === "string" || Elixir.Core.is_instance_of(unquote(term), String);
    end
  end

  defmacro is_bitstring(term) do
    quote do
      Elixir.Core.is_instance_of(unquote(term), Elixir.Core.BitString);
    end
  end

  defmacro is_boolean(term) do
    quote do
      typeof(unquote(term)) === "boolean" || Elixir.Core.is_instance_of(unquote(term), Boolean);
    end
  end

  defmacro is_float(term) do
    quote do
      is_number(unquote(term)) && !Number.isInteger(unquote(term))
    end
  end

  defmacro is_function(term) do
    quote do
      typeof(unquote(term)) === "function" || Elixir.Core.is_instance_of(unquote(term), Function);
    end
  end

  defmacro is_function(term, _arity) do
    quote do
      typeof(unquote(term)) === "function" || Elixir.Core.is_instance_of(unquote(term), Function);
    end
  end

  defmacro is_integer(term) do
    quote do
      Number.isInteger(unquote(term))
    end
  end

  defmacro is_list(term) do
    quote do
      Array.is_array(unquote(term))
    end
  end

  defmacro is_number(term) do
    quote do
      is_integer(unquote(term)) || is_float(unquote(term))
    end
  end

  defmacro is_pid(term) do
    quote do
      Elixir.Core.is_instance_of(unquote(term), Elixir.Core.PID);
    end
  end

  defmacro is_tuple(term) do
    quote do
      Elixir.Core.is_instance_of(unquote(term), Elixir.Core.Tuple);
    end
  end

  defmacro is_map(term) do
    quote do
      typeof(unquote(term)) === "object" || Elixir.Core.is_instance_of(unquote(term), Object);
    end
  end

  defmacro length(term) do
    quote do
      unquote(term).length
    end
  end

  defmacro map_size(term) do
    quote do
      Object.keys(unquote(term)).length
    end
  end

  def max(first, second) do
    Math.max(first, second)
  end

  def min(first, second) do
    Math.min(first, second)
  end

  def round(number) do
    Math.round(number)
  end

  def trunc(number) do
    Math.floor(number)
  end

  def tuple_size(tuple) do
    tuple.length
  end

  def not(arg) do
    !arg
  end

  def elem(tuple, index) do
    tuple.get(index)
  end

  defmacro is_nil(term) do
    quote do
      unquote(term) == nil
    end
  end

  defmacro match?(left, right) do
    quote do
      case unquote(right) do
        unquote(left) ->
          true
        _ ->
          false
      end
    end
  end



end
